### ISSUE TITLE
Implement Conditions

### ISSUE THREAD

**Issue by pydctw** [kubernetes-sigs/cluster-api-provider-gcp#570]
/kind feature

**Describe the solution you'd like**
Add Conditions to `GCPCluster.Status` and `GCPMachine.Status`

CAPI ecosystem implemented Conditions to show the Cluster status at glance. It will be great to add conditions to GCPCluster and GCPMachine to help users to understand the status of each object better.

Tasks
* https://github.com/kubernetes-sigs/cluster-api-provider-gcp/issues/577
* https://github.com/kubernetes-sigs/cluster-api-provider-gcp/issues/576

These tasks can be done in parallel. Feel free to pick one or the other.

**Related docs**
* k8s API conventions: https://github.com/kubernetes/community/blob/a2cdce5/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties
* CAEP: https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20200506-conditions.md

**Anything else you would like to add:**
Other providers have implemented conditions so take a look at CAPA or CAPZ for reference.

**Comment by s1ntaxe770r** [comment]
Hi, @pydctw I would like to work on this. I did some research using some of the links you provided and looking at CAPZ and here is what I understand so far: 

A few new types would need to be introduced like CAPZ does over here:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/82f169f652ecf26dee9f789b17bbd86776a02f88/api/v1beta1/conditions_consts.go#L23

The current GCPmachine status spec would need to be updated to include  `clusterv1.Conditions`  field 

Here's what I currently don't understand 

- The exact types of conditions is there some kind of documentation for the possible state of a gcp machine ? 

- How I'd implement this in the reconciliation loop of the GCP machine controller? ( I suppose I should take a look at CAPA )

I'm probably missing a lot more but I hope this is sufficient information to get started

**Comment by pydctw** [comment]
Hi @s1ntaxe770r, thanks for your interest in the issue. Very happy to hear that.

You are on the correct path -  providers usually create a file called conditions_consts.go to define the condition types and clusterv1.Conditions field needs to be added to GCPmachine status.

> The exact types of conditions is there some kind of documentation for the possible state of a gcp machine ?

Each providers have slightly different conditions. For GCP, take a look at what's returned from GCP instance get call
* https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/main/cloud/services/compute/instances/reconcile.go#L130
* https://pkg.go.dev/google.golang.org/api@v0.73.0/compute/v1#Instance - Instance has a Status field

> How I'd implement this in the reconciliation loop of the GCP machine controller? ( I suppose I should take a look at CAPA )

Yes, reading other providers code will definitely give you a good idea as infraMachine controllers has same structure. 

I would suggest reading [awsmachine_controller](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/controllers/awsmachine_controller.go) and investigate [conditions](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/eaafc000e33e37d16b1cd1b90795674f3fd8d521/pkg/cloud/scope/machine.go#L307-L315).

**Comment by s1ntaxe770r** [comment]
Hey, @pydctw  thanks for pointing me in the right direction. I've looked through the links and have a few more questions 

-  I noticed CAPA and CAPZ have a concept of "Reasons" example over [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/eaafc000e33e37d16b1cd1b90795674f3fd8d521/api/v1beta1/conditions_consts.go#L123) would this be necessary for GCP machines?  

- Is there any benefit in patching the status of the machine vs Fetching the status of the machine by calling `instances.get()`

- Is this a good time to split this issue into two parts? I would like to start with GCPmachine first
